packages/remote-feature-flag-controller/src/utils/user-segmentation-utils.ts

```diff
@@ -1,25 +1,49 @@
 import type { Json } from '@metamask/utils';
+import { validate as uuidValidate, version as uuidVersion } from 'uuid';
 
 import type { FeatureFlagScopeValue } from '../remote-feature-flag-controller-types';
 
-/* eslint-disable no-bitwise */
 /**
  * Generates a deterministic random number between 0 and 1 based on a metaMetricsId.
  * This is useful for A/B testing and feature flag rollouts where we want
  * consistent group assignment for the same user.
  *
- * @param metaMetricsId - The unique identifier used to generate the deterministic random number
+ * Supports two metaMetricsId formats:
+ * - UUIDv4 format (new mobile implementation)
+ * - Hex format with 0x prefix (extension or old mobile implementation)
+ *
+ * For UUIDv4 format, the following normalizations are applied:
+ * - Replaces version (4) bits with 'f' to normalize range
+ * - Replaces variant bits (8-b) with 'f' to normalize range
+ * - Removes all dashes from the UUID
+ *
+ * For hex format:
+ * - Expects a hex string with '0x' prefix (e.g., '0x1234abcd')
+ * - Removes the '0x' prefix before conversion
+ * - Converts the remaining hex string to a BigInt for calculation
+ *
+ * @param metaMetricsId - The unique identifier used to generate the deterministic random number, can be a UUIDv4 or hex string
  * @returns A number between 0 and 1 that is deterministic for the given metaMetricsId
  */
 export function generateDeterministicRandomNumber(
   metaMetricsId: string,
 ): number {
-  const hash = [...metaMetricsId].reduce((acc, char) => {
-    const chr = char.charCodeAt(0);
-    return ((acc << 5) - acc + chr) | 0;
-  }, 0);
-
-  return (hash >>> 0) / 0xffffffff;
+  let cleanId: string, value: bigint;
+  // uuidv4 format
+  if (uuidValidate(metaMetricsId) && uuidVersion(metaMetricsId) === 4) {
+    cleanId = metaMetricsId
```

Comment:
Just a thought: perhaps it would be easier to understand if we did this all in one step, e.g.

```suggestion
    result = metaMetricsId.match(/^([0-9a-f]{8})-([0-9a-f]{4})-4([0-9a-f]{3})-([0-9a-f]{4})-([0-9a-f]{12})$/)
    if (result === null) {
      throw new Error('Failed to parse UUID')
    }
    const [_completeMatch, ...captureGroups] = result;
    cleanId = captureGroups.join('');
```

It's a bit confusing to read, but avoids the need to count digits (because it will fail to match if we got the counts wrong, validating the positions for us).

We could also break up the Regex more by using a Regex constructer rather than a literal, to make it more readable.

Reply:
Regarding this one, I wanted to take advantage of validation from `uuidv4` library => `uuidValidate` and `uuidVersion` so we don't need to perform regex match for validation. I added some explanation in jsdoc to make it easier to read 🙏🏻 let me know your thoughts

Reply:
Agreed that we should use the `validate` function, the error case in my suggestion would not be for invalid inputs. It would be as a guard in case of programmer error (e.g. in case our regex is wrong), hopefully unreachable by a test, so we can skip it from coverage checks.

The current solution would not warn us if the regex was wrong (i.e. if the `replace` had no affect because the index was wrong). That's the advantage this would provide.

Just to explain my reasoning, I don't think I want to block on this though.

Reply:
It's always great to improve the code readibility and functionality ! thanks for the suggestions 🫶 It is not blocking i'd like to ship something solid and without going back or revisiting. Validation is added 3ae4dd85c499f9f05d40d84425ad6235757493dd